ada/language/ast.py

```diff
@@ -7317,31 +7342,74 @@ def no_overload_equation():
 
 
             lambda _: Or(
-                # Regular case: Both operands and binop are of the same type.
-                # We call base_subtype as a conversion property because
-                # operators are defined on the root subtype, so the return
-                # value will always be of the root subtype.
-                #
-                # TODO: inlining of the TypeBind macro.
-                Bind(Self.left.type_var, Self.type_var,
-                     eq_prop=BaseTypeDecl.matching_type,
-                     conv_prop=BaseTypeDecl.base_subtype_or_null)
+                # TODO: several inlinings of the TypeBind macro below.
+                # TODO: the 4 cases below can be made into a single equation
+                # once we have conversion properties that can take multiple
+                # logic variables as parameters.
+
+                # note: in the relevant cases below, we call base_subtype as a
+                # conversion property because operators are defined on the root
+                # subtype, so the return value will always be of the root
+                # subtype.
+
+                # Case 1: Both operands have an universal type: set the result
+                # type to that universal type.
+                Predicate(BaseTypeDecl.is_universal_type,
```

Comment:
That's really dirty. I think this should be the last patch of this kind until we merge the new solver ...